dbms/src/Flash/Mpp/GRPCSendQueue.h

```diff
@@ -0,0 +1,285 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Exception.h>
+#include <Common/Logger.h>
+#include <Common/MPMCQueue.h>
+#include <Common/grpcpp.h>
+#include <common/logger_useful.h>
+
+#include <functional>
+
+namespace DB
+{
+
+namespace tests
+{
+class TestGRPCSendQueue;
+} // namespace tests
+
+/// In grpc cpp framework, the tag that is pushed into grpc completion
+/// queue must be inherited from `CompletionQueueTag`.
+class KickTag : public ::grpc::internal::CompletionQueueTag
+{
+public:
+    enum class Status
+    {
+        // No tag.
+        NONE,
+        // Waiting for kicking.
+        WAITING,
+        // Queuing in the grpc completion queue.
+        QUEUING,
+    };
+
+    KickTag(std::mutex * m, const LoggerPtr & l)
+        : mu(m)
+        , log(l)
+    {}
+
+    ~KickTag() override
+    {
+        std::unique_lock lock(*mu);
+
+        RUNTIME_ASSERT(status == Status::NONE, log, "status {} is not none", status);
+    }
+
+    bool FinalizeResult(void ** tag_, bool * /*status*/) override
+    {
+        std::unique_lock lock(*mu);
+
+        RUNTIME_ASSERT(status == Status::QUEUING, log, "status {} is not queuing", status);
+        status = Status::NONE;
+
+        *tag_ = tag;
+        tag = nullptr;
+
+        return true;
+    }
+
+    Status getStatus()
+    {
+        std::unique_lock lock(*mu);
+        return status;
+    }
+
+private:
+    friend class tests::TestGRPCSendQueue;
+    template <typename T>
+    friend class GRPCSendQueue;
+
+    /// Should acquire lock.
+    bool toWaiting(void * new_tag)
```

Comment:
Still worry about the unclear ownership of mu.

A way to avoid depending on the caller locking mutex is to pass in the logic that should call in the critical area:

```cpp
template <typename F>
auto toWaiting(F && precheck, void * new_tag) {
    std::unique_lock lock(mu); // KickTag itself could own the mutex now.
    auto status = precheck();
    if (status == GRPCSendQueueRes::EMPTY) {
        ...
    }
    return status;
}
```

Reply:
It's not very elegant.
I think a while and have changed it to a more reasonable manner(in my opinion).
The `status` and `tag` should belong to `GRPCSendQueue`. A `std::function` that contains the mutex logic is passed to `KickTag`. In this way, there is only one mutex.